File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
class Solution {
    boolean solution(String s) {
        int openCount = 0;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (c == '(') {
                openCount++;
            } else { // ')'
                openCount--;

                // 닫는 괄호가 더 많아지는 순간 바로 실패
                if (openCount < 0) {
                    return false;
                }
            }
        }

        // 열린 괄호가 남아있으면 실패
        return openCount == 0;
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    answer = True
    s_list = list(s)
    l_list = []
    a = ""
    for b in s_list:
        a = b
        if a == '(':
            l_list.append(a)
        elif len(l_list) == 0:
            return False
        else:
            l_list.pop()
    
    if (a == '(' or len(l_list) != 0):
        return False
    return True